tanglegram/tangle.py

```diff
@@ -171,73 +167,148 @@ def gen_tangle(a, b, labelsA=None, labelsB=None, optimize_order=10000,
                         linkages to be matched
     (labels1,labels2) : list of str
-
-
     Returns
     -------
     optimized linkage
     """
 
-    dend1 = sclust.hierarchy.dendrogram(link1, orientation = 'left',
-                                        no_plot=True, labels=labels1)
-    dend2 = sclust.hierarchy.dendrogram(link2, orientation = 'right',
-                                        no_plot=True, labels=labels2)
-
-    entangles =[get_entanglement(dend1, dend2)]
-
-    pbar = trange(max_iter)
-
-    for i in pbar:
-        temp_link1 = link1.copy()
-        temp_link2 = link2.copy()
-
-        # Go over all "hinges" of the dendrogram
-        for k in range(len(link2)):
-            # In 50% of the cases turn the hinge
-            if random.randint(0, 100) > 50:
-                temp_link2[k] = [temp_link2[k][1], temp_link2[k][0], temp_link2[k][2], temp_link2[k][3]]
-
-        # Do the same for other dendrogram
-        for k in range(len(link1)):
-            if random.randint(0,100) > 50:
-                temp_link1[k] = [temp_link1[k][1], temp_link1[k][0], temp_link1[k][2], temp_link1[k][3]]
-
-        # Generate actual dendrogram to get leaf order
-        dend1 = sclust.hierarchy.dendrogram(temp_link1, orientation = 'left',
-                                            no_plot=True, labels=labels1)
-        dend2 = sclust.hierarchy.dendrogram(temp_link2, orientation = 'right',
-                                            no_plot=True, labels=labels2)
-
-        # Test new entanglement -> keep if better then previous
-        new_entangle = get_entanglement(dend1, dend2)
-        if entangles[-1] >= new_entangle:
-            entangles.append(new_entangle)
-            link1 = temp_link1
-            link2 = temp_link2
-        else:
-            entangles.append( entangles[-1] )
-
-        if i > 0 and i % 100 == 0:
-            slope = round(entangles[-1] - np.mean(entangles[-100:-1]), 2)
-            pbar.set_description('Optimising {0}'.format(slope))
-
-    pbar.close()
-
-    module_logger.info('Finished optimising at entanglement {0}'.format(round(entangles[-1], 4)))
-
-
-    return link1,link2
+    old_entang = -1
+    cond = 1
+    final_entang = 10
+    save_entang = []
+
+    for li_MID in range(len(linkage_matrix1) - 1, 1, -1):
+        all_linkage1 = get_all_linkage(linkage_matrix1, li_MID)
+        all_linkage2 = get_all_linkage(linkage_matrix2, li_MID)
+
+        for i in all_linkage1:
+            for j in all_linkage2:
+                best_linkage1 = i
+                best_linkage2 = j
+                dend1 = sclust.hierarchy.dendrogram(best_linkage1, labels=labels1)
+                dend2 = sclust.hierarchy.dendrogram(best_linkage2, labels=labels2)
+                min_entang = get_entanglement(dend1, dend2)
+                old_rem = -1
+                if len(save_entang) == 0:
+                    save_entang.append(min_entang)
+                for k in range(10):
```

Comment:
Just having a quick glance - will have a thorough look later but where is this hardcoded `range(10)` coming from?

Reply:
I can change the convergence condition to make it more sensible